stacks-signer/src/crypto/frost.rs

```diff
@@ -0,0 +1,780 @@
+            sign_id: self.current_sign_id,
+            sign_nonce_id: self.current_sign_nonce_id,
+        };
+        let nonce_request_msg = Message {
+            sig: nonce_request.sign(&self.message_private_key).expect(""),
+            msg: MessageTypes::NonceRequest(nonce_request),
+        };
+        self.ids_to_await = (0..self.total_signers).collect();
+        self.move_to(State::NonceGather)?;
+        Ok(nonce_request_msg)
+    }
+
+    fn gather_nonces(&mut self, message: &Message) -> Result<(), Error> {
+        match &message.msg {
+            MessageTypes::NonceResponse(nonce_response) => {
+                // TODO: check sign_id and sign_nonce_id
+                self.public_nonces
+                    .insert(nonce_response.signer_id, nonce_response.clone());
+                self.ids_to_await.remove(&nonce_response.signer_id);
+                info!(
+                    "Sign round #{} nonce round #{} NonceResponse from signer #{}. Waiting on {:?}",
+                    nonce_response.sign_id, nonce_response.sign_nonce_id, nonce_response.signer_id, self.ids_to_await
+                );
+            }
+            _ => {}
+        }
+        if self.ids_to_await.is_empty() {
+            // Calculate the aggregate nonce
+            let aggregate_nonce = self
+                .dkg_public_shares
+                .iter()
+                .fold(Point::default(), |s, (_, dps)| s + dps.public_share.A[0]);
+            // check to see if aggregate public key has even y
+            if aggregate_nonce.has_even_y() {
+                info!("Aggregate nonce has even y coord!");
+                info!("Aggregate nonce: {}", aggregate_nonce);
+                self.move_to(State::SigShareRequest)?;
+            } else {
+                warn!("Sign Round #{} Nonce Round #{} Failed: Aggregate nonce does not have even y coord, requesting new nonces.", self.current_sign_id, self.current_sign_nonce_id);
+                self.current_sign_nonce_id += 1;
+                self.move_to(State::NonceRequest)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn request_sig_shares(&mut self) -> Result<Message, Error> {
+        info!(
+            "Sign Round #{} Requesting Signature Shares",
+            self.current_sign_id, 
+        );
+        let nonce_responses = (0..self.total_signers).map(|i| self.public_nonces[&i].clone()).collect::<Vec<NonceResponse>>();
+        let sig_share_request = SignatureShareRequest {
+            dkg_id: self.current_dkg_id,
+            sign_id: self.current_sign_id,
+            correlation_id: 0, // TODO: what is this?
+            nonce_responses,
+            message: self.message.clone(),
+        };
+        let sig_share_request_msg = Message {
+            sig: sig_share_request.sign(&self.message_private_key).expect(""),
+            msg: MessageTypes::SignShareRequest(sig_share_request), // TODO: either SigShare or SignatureShare, SignShare is right out
```

Comment:
Will this TODO be addressed before merge?

Reply:
Can't do it as part of this PR because the relevant code is in `stacks-sbtc`.  We are planning on number of `stacks-sbtc` changes before the follow on PR.

Reply:
Gotcha, thanks!